crates/uv-auth/src/middleware.rs

```diff
@@ -100,88 +128,184 @@ impl Middleware for AuthMiddleware {
+                self.complete_request(Some(credentials), request, extensions, next)
+                    .await
+            } else {
+                trace!("Request for {url} is missing a password, looking for credentials");
+                // There's just a username, try to find a password
+                let credentials = if let Some(credentials) = self
+                    .cache()
+                    .get_realm(NetLoc::from(request.url()), credentials.to_username())
+                {
+                    request = credentials.authenticate(request);
+                    // Do not insert already-cached credentials
+                    None
+                } else if let Some(credentials) = self
+                    .fetch_credentials(Some(&credentials), request.url())
+                    .await
+                {
+                    request = credentials.authenticate(request);
+                    Some(Arc::new(credentials))
+                } else {
+                    // If we don't find a password, we'll still attempt the request with the existing credentials
+                    Some(credentials)
+                };
+
+                self.complete_request(credentials, request, extensions, next)
+                    .await
             }
-        // Otherwise, look for complete credentials in:
-        // (3) The netrc file
-        } else if let Some(credentials) = self.netrc.as_ref().and_then(|netrc| {
+        } else {
+            // We have no credentials
+            trace!("Request for {url} is unauthenticated, checking cache");
+
+            // Check the cache for a URL match
+            let credentials = self.cache().get_url(request.url(), Username::none());
+            if let Some(credentials) = credentials.as_ref() {
+                request = credentials.authenticate(request);
+                if credentials.password().is_some() {
+                    return self.complete_request(None, request, extensions, next).await;
+                }
+            }
+
+            // Otherise, attempt an anonymous request
+            trace!("Attempting unauthenticated request for {url}");
+
+            // <https://github.com/TrueLayer/reqwest-middleware/blob/abdf1844c37092d323683c2396b7eefda1418d3c/reqwest-retry/src/middleware.rs#L141-L149>
+            // Clone the request so we can retry it on authentication failure
+            let mut retry_request = request.try_clone().ok_or_else(|| {
+                Error::Middleware(anyhow!(
+                    "Request object is not clonable. Are you passing a streaming body?".to_string()
+                ))
+            })?;
+
+            let response = next.clone().run(request, extensions).await?;
+
+            // And on failure, search for credentials
+            if matches!(
+                response.status(),
+                StatusCode::FORBIDDEN | StatusCode::NOT_FOUND | StatusCode::UNAUTHORIZED
+            ) {
```

Comment:
In what cases will we need to do this? Like, when would we expect this to happen, and when would we expect that we can retrieve from the cache?

Reply:
Maybe to put it differently: when would we expect to see cache misses for URLs on the same domain?

Reply:
If you have `example.com/foo` and it provides a link to `example.com/bar` we will not match in the URL-level cache and must attempt an unauthorized request first because `bar` may not need credentials and could fail if provided. On failure, we'll find credentials in the realm-level cache and try them.